<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Описание-данных-" data-toc-modified-id="Описание-данных--1"><span class="toc-item-num">1&nbsp;&nbsp;</span><strong><font color="DarkOliveGreen">Описание данных </font></strong></a></span></li><li><span><a href="#-Откроем-файл-с-данными-и-изучим-общую-информацию---" data-toc-modified-id="-Откроем-файл-с-данными-и-изучим-общую-информацию----2"><span class="toc-item-num">2&nbsp;&nbsp;</span><strong><font color="Olive"> Откроем файл с данными и изучим общую информацию </font></strong>  <a id="2" rel="nofollow"></a></a></span><ul class="toc-item"><li><span><a href="#-Общая-информация-о-данных:-" data-toc-modified-id="-Общая-информация-о-данных:--2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span><strong><font color="OliveDrab"> Общая информация о данных: </font></strong></a></span></li></ul></li><li><span><a href="#Предобработка-данных-" data-toc-modified-id="Предобработка-данных--3"><span class="toc-item-num">3&nbsp;&nbsp;</span><strong><font color="YellowGreen">Предобработка данных </font></strong></a></span><ul class="toc-item"><li><span><a href="#--Обработка-пропусков-" data-toc-modified-id="--Обработка-пропусков--3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span><strong><font color="SeaGreen">  Обработка пропусков </font></strong></a></span></li><li><span><a href="#-Замена-типа-данных-" data-toc-modified-id="-Замена-типа-данных--3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span><strong><font color="DarkSeaGreen"> Замена типа данных </font></strong></a></span></li><li><span><a href="#-Обработка-дубликатов-" data-toc-modified-id="-Обработка-дубликатов--3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span><strong><font color="LightSeaGreen"> Обработка дубликатов </font></strong></a></span></li><li><span><a href="#-Лемматизация-" data-toc-modified-id="-Лемматизация--3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span><strong><font color="DarkCyan"> Лемматизация </font></strong></a></span></li><li><span><a href="#-Категоризация-данных-" data-toc-modified-id="-Категоризация-данных--3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span><strong><font color="DodgerBlue"> Категоризация данных </font></strong></a></span></li></ul></li><li><span><a href="#-Ответы-на-вопросы-" data-toc-modified-id="-Ответы-на-вопросы--4"><span class="toc-item-num">4&nbsp;&nbsp;</span><strong><font color="CornflowerBlue"> Ответы на вопросы </font></strong></a></span></li><li><span><a href="#-Общий-вывод---" data-toc-modified-id="-Общий-вывод----5"><span class="toc-item-num">5&nbsp;&nbsp;</span><strong><font color="MediumSlateBlue"> Общий вывод </font></strong>  <a id="4" rel="nofollow"></a></a></span></li><li><span><a href="#-Чек-лист-готовности-проекта-" data-toc-modified-id="-Чек-лист-готовности-проекта--6"><span class="toc-item-num">6&nbsp;&nbsp;</span><strong><font color="RoyalBlue"> Чек-лист готовности проекта </font></strong></a></span></li></ul></div>

**ПРЕДОБРАБОТКА ДАННЫХ**

# Исследование надёжности заёмщиков  <a id='1'></a>

**<font color='green'> Заказчик </font>**  — кредитный отдел банка.   

**<font color='red'> Цель исследования</font>**  - **узнать, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.**

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

## **<font color='DarkOliveGreen'>Описание данных </font>**  ###

**children** — количество детей в семье  
**days_employed** — общий трудовой стаж в днях  
**dob_years** — возраст клиента в годах  
**education** — уровень образования клиента  
**education_id** — идентификатор уровня образования  
**family_status** — семейное положение  
**family_status_id** — идентификатор семейного положения  
**gender** — пол клиента  
**income_type** — тип занятости  
**debt** — имел ли задолженность по возврату кредитов  
**total_income** — ежемесячный доход  
**purpose** — цель получения кредита

**<font color='MediumSeaGreen'> Краткий план работы:</font>**  

1. Изучить данные
3. Обработать пропуски. Посмотреть, какие значения у нас пропущены (не заполнены), по какой причине (являются ли пропуски случайными или нет). Заполнить недостающие значения медианами там, где это нужно. 
4. Заменить некорректные данные на корректные.
5. Обработать дубликаты. Сначала посмотреть списки уникальных значений, чтобы из них понять, где могут быть неявные дубликаты. Затем привести к единому регистру для того, чтобы неявные дубликаты стали явными. Затем удалить все явные дубликаты.
5. Лемматизировать (т.е. привести к единой словарной форме) цели кредита. 
6. Категоризировать данные по семейному положению, количеству детей, уровню дохода (низкий, средний, высокий) и целям кредита (здесь категоризация производится на основе результатов лемматизации).
7. сделать выводы на основе категоризации данных и статистики из сводных таблиц. Выводы о том, какие категории заемщиков наиболее ответственные, а какие - наименее. И, конечно же, вывод о том, кореллирует ли семейное положение с количеством просрочек, а также наличие детей с количеством просрочек.

## **<font color='Olive'> Откроем файл с данными и изучим общую информацию </font>**  <a id='2'></a>

In [1]:
import pandas as pd
from pymystem3 import Mystem
from nltk.stem import SnowballStemmer 
from collections import Counter

In [2]:
#добавим красоты в оформление проекта 

from IPython.core.display import display, HTML

display(HTML("<font color='blue'>синий текст</font>"))
display(HTML("<font color='maroon'>коричневый текст</font>"))
display(HTML("<font color='olive'>текст оливкогового цвета</font>"))
display(HTML("<font color='navy'>темно-синий текст</font>"))
display(HTML("<font color='purple'> фиолетовый текст</font>"))
display(HTML("<font color='teal'>бирюзовый текст</font>"))
display(HTML("<font color='gray'>серый текст</font>"))
display(HTML("<font color='lime'>светло-зеленый текст</font>"))
display(HTML("<font color='green'>зеленый текст</font>"))
display(HTML("<font color='yellow'>желтый текст</font>"))
display(HTML("<font color='FUCHSIA'> розовый текст</font>"))
display(HTML("<font color='AQUA'> голубой текст</font>"))

In [3]:
df = pd.read_csv('/datasets/data.csv') 

###  **<font color='OliveDrab'> Общая информация о данных: </font>**  ###

In [4]:
#выведем на экран первые 10 строк датафрейма для того, чтобы увидеть структуру данных и другие особенности

df.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

**Заметки**

1. В столбце **"education"** есть одинаковые значения, напечатанные разными регистрами. Например, "среднее" и "СРЕДНЕЕ".
Это неявные дубликаты. Для того, чтобы их удалить, приведем их в дальнейшем к нижнему регистру для легкого обнаружения дубликатов.
2. В столбце с количеством дней трудового стажа **("days_employed")** есть отрицательные значения. Возможно, это тоже опечатка. Возьмем эти значения по модулю. 
3. В столбце **"days_employed"** данные неоднородны - в каких-то ячейках количество указано в днях, в каких-то - в часах. 340266 (см.строку с индексом 4) - это, скорее всего, количество часов. Но точно не количество дней. Исправим, переведя часы в дни. 
4. Значения столбца **"family_status_id"** соответствуют значениям столбца **"family_status"**. 0 -женат/замужем, 1 - гражданский брак, 2 - вдовец/вдова и т.д.
5. Есть пропуски в столбцах **"total_income"** и **"days_employed"**.


In [4]:
#общая информация о датафрейме

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [5]:
df.columns

Index(['children', 'days_employed', 'dob_years', 'education', 'education_id',
       'family_status', 'family_status_id', 'gender', 'income_type', 'debt',
       'total_income', 'purpose'],
      dtype='object')

**Заметки**

1. Не все строки из столбцов **"days_employed"** и **"total_income"** заполнены. Число пропусков в обоих столбцах одинаково.

21525 (заполненные значения во всех столбцах) - 19351 (заполненные значения в столбцах "days_employed" и "total_income") = 2174 - это количество пропусков в вышеупомянутых столбцах.

2. Здесь мы видим, какие типы данных (int, float, object) есть в датафрейме

In [6]:
# с помощью атрибута shape мы можем узнать количество строк (первая цифра) и столбцов (вторая цифра) в датафрейме

df.shape

(21525, 12)

В нашем датафрейме есть 21525 строк и 12 столбцов

In [7]:
# типы данных

df.dtypes

children              int64
days_employed       float64
dob_years             int64
education            object
education_id          int64
family_status        object
family_status_id      int64
gender               object
income_type          object
debt                  int64
total_income        float64
purpose              object
dtype: object

При помощи атрибута **dtypes** мы видим, какие типы данных есть в нашем датафрейме.

**В нашем датафрейме встречаются данные трех типов:**  

**object** - текст  
**float64** -  дробное число (также известное как "число с плавающей точкой", вещественный тип данных)   
**int64** - целое число (оно же - целочисленный тип данных)

In [8]:
# описание данных 
# пользуясь методом describe, смотрим на описательную статистику числовых данных:
# минимальные (min) и максимальные (max) значения столбцов, среднее (mean), медиана (median), стандратное отклонение (std)
# и процентили (25%, 50%, 75%). 

df.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

In [9]:
df.index

RangeIndex(start=0, stop=21525, step=1)

In [10]:
#переводим среднее значение дней трудового стажа из дней в годы

df['days_employed'].mean() / 365

172.73013057937914

In [11]:
# переводим медиану дней трудового стажа из дней в годы

df['days_employed'].median() / 365

-3.296902818549285

In [12]:
# смотрим стандартное отклонение

df['days_employed'].std() / 365

385.8282519844296

In [13]:
#считаем количество семей с 20 детьми

print("Количество семей с 20 детьми:", df[df['children'] == 20].count()[0])

Количество семей с 20 детьми: 76


In [14]:
#удаляем строки с семьями, где детей 20

df.drop(df.loc[df['children']==20].index, inplace=True)

In [15]:
#проверяем, удалились ли эти строки

print("Количество семей с 20 детьми:", df[df['children'] == 20].count()[0])

Количество семей с 20 детьми: 0


In [16]:
#заменим количество детей -1 на 0 для единообразия в датафрейме

df['children'] = df['children'].replace(-1, 0)

In [17]:
print("Количество семей с 0 детьми:", df[df['children'] == 0].count()[0])

Количество семей с 0 детьми: 14196


In [18]:
#проверяем, заменились ли строки с количеством детей "-1" на "0"

print("Количество семей с -1 детьми:", df[df['children'] == -1].count()[0])

Количество семей с -1 детьми: 0


In [19]:
#проверим, нет ли в базе значений столбца 'dob_years' (возраст) со значением 0

print("Количество людей с нулевым возрастом:", df[df['dob_years'] == 0].count()[0])

Количество людей с нулевым возрастом: 100


In [20]:
#смотрим уникальные значения по столбцу "пол клиента"

df['gender'].value_counts()

F      14189
M       7259
XNA        1
Name: gender, dtype: int64

XNA - это у нас кто, гермафродит? Гугл не дал мне ответ на этот вопрос. Думаю, можно удалить.

In [21]:
#удалим гермафродита

df.drop(df.loc[df['gender']== 'XNA'].index, inplace=True)

In [22]:
#проверим, удалили ли

df['gender'].value_counts()

F    14189
M     7259
Name: gender, dtype: int64

**Заметки**

1. **<font color='green'> Минимальное количество детей </font>**   (столбец **"Children"**) указано -1.   
Преподаватель ответил, что это 0 и -1 означает отсутствие детей. А получилось так потому, что данные выгружали из нескольких разных систем и в них был разный формат записи.  

2. **<font color='green'> Максимальное количество детей </font>**  - 20. С большой долей вероятности это опечатка. Цифры 2 и 0 - это соседние клавиши на цифровом блоке клавиатуры, поэтому у человека, вводившего данные, палец соскользнул с одной клавиши на другую. Это надо будет исправить.  

3. В столбце **<font color='green'> с возрастом клиентов "dob_years" есть значения, равные нулю </font>**    - предполагаю, что это потому, что возраст не был указан изначально. Отстутсвующие значения.  

4. **<font color='green'> Среднее значение в столбце **days_employed (кол-во дней трудового стажа)** </font>** , если перевести его из дней в годы, разделив среднее значение на 365, показало  **<font color='green'> 172 года </font>** . Такого не может быть. Как я уже писала выше, выгрузка данных была из разных систем, где в одних системах данные о трудовом стаже были указаны в днях, а в других - в часах. Надо будет привести данные к одному формату (перевести часы в дни).  

5. **<font color='green'> текст </font>** Помимо среднего значения по **days_employed**, я еще в отдельных ячейках посмотрела по этому же столбцу медиану, стандратное отклонение и разделила на 365, переведя дни в годы. Что касается медианы - я думаю, что медиана в 5 лет - вполне реальная цифра. Вот только насчет стандартного отклонения в 14 лет не понятно, нормально ли это.   

6. **<font color='green'> текст </font>** Количество семей с 20 детьми: 76. Я  думаю, что все это - опечатки потому, что клавиши 2 и 0 на цифровом блоке - соседние и 
человек, вводивший данные, мог, нажимая 2 или 0, случайно задеть соседнюю клавишу. Чем заменить эти данные, не понятно (количеством детей 2 или 0, не понятно).  Но поскольку этих данных меньше 1%, то я их удалила.  

7. Обнаружено 100 **<font color='green'> клиентов с нулевым возрастом. </font>**   Такого тоже не может быть. Это явно незаполненные значения. С этим надо что-то делать.   

8. **<font color='green'> текст </font>** В принципе, можно было бы некорретными значениями счесть и те строки, где трудовой стаж нулевой. Но я погуглила и узнала, что банки ввиду высокой конкуренции меняют условия и становятся более лояльными к клиентам. Некоторые из них выдают кредиты и тем, у кого стажа совсем мало, либо его нет вообще. Поэтому эти строки трогать не будем.


**Вывод**

В датафрейме были обнаружены некорректные данные, образовавшиеся в результате опечаток, а также в в результате выгрузки из разных систем данных с разными форматами записей. Часть из них отредактированы для создания единообразия в датафрейме, часть - удалены.

## **<font color='YellowGreen'>Предобработка данных </font>** 

### **<font color='SeaGreen'>  Обработка пропусков </font>**

***Задание:***

1. Определите и заполните пропущенные значения.
2. Опишите, какие пропущенные значения вы обнаружили;
3. Приведите возможные причины появления пропусков в данных;
4. Объясните, по какому принципу заполнены пропуски;

***Пояснения:***

1. **df.isna().sum()** -  подсчёт пропущенных значений. 

2. В **days_employed** обнаружила **NaN** (вывела первые 5 строк. Пробовала выводить и 20, и 30, чтобы убедиться наверняка - ничего не меняется. Кроме NaN, больше ничего не выводит среди типов пропущенных значений.

3. Совпадает количество пропусков в столбцах **days_employed** и ****total_income** -2174. Это чуть меньше 10% от общего числа заемщиков.
Возможно, часть замещиков брали кредиты, не предоставляя справку о доходах и копию трудовой книжки.
Поэтому сотрудники банка не заполнили сведения о сумме всех источников дохода (total_income) и трудовом стаже(**days_employed**) 

4. Думаю, в таком случае нужно заполнить пропуски медианными значениями зарплат и кол-ва дней трудового стажа

In [23]:
# считаем пропуски в столбце days_employed (пропуски в трудовом стаже)

display('Пропуски в трудовом стаже до:', df['days_employed'].isna().sum())
display(df[df['days_employed'].isna()].head()) 

'Пропуски в трудовом стаже до:'

2165

children  days_employed  dob_years education  education_id  \
12         0            NaN         65   среднее             1   
26         0            NaN         41   среднее             1   
29         0            NaN         63   среднее             1   
41         0            NaN         50   среднее             1   
55         0            NaN         54   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   

    total_income                           purpose  
12           NaN                   сыграть свадьбу  
26           NaN                       образование  
29           NaN  строительство жилой недвижимости  
41           NaN  сделка с подержанным автомобилем  
55           NaN                   сыграть свадьбу

In [24]:
#переводим среднее значение дней трудового стажа из дней в годы

df['days_employed'].mean() / 365

172.99898710353676

In [25]:
# выводим на экран значения до получения модуля дней трудового стажа. Среди них есть отрицательные.
# такого не может быть потому, что у значений "дни трудового стажа" не может быть отрицательного значения
# значения могут быть либо нулевые (человек нигде никогда не работал, поэтому у него дней трудового стажа ноль)
# либо положительные (человек работал и его трудовой стаж можно измерить в днях)

display('первые 5 строк до получения модуля дней трудового стажа', df['days_employed'].head())

'первые 5 строк до получения модуля дней трудового стажа'

0     -8437.673028
1     -4024.803754
2     -5623.422610
3     -4124.747207
4    340266.072047
Name: days_employed, dtype: float64

In [26]:
# берем значения по модулю

df['days_employed'] = df['days_employed'].apply(abs) 
hours_to_days = df.loc[df['days_employed'] > 20000, 'days_employed'] = df.loc[df['days_employed'] >20000, 'days_employed'] / 24

display('Количество дней трудового стажа после перевода часов в дни', hours_to_days.head())

'Количество дней трудового стажа после перевода часов в дни'

4     14177.753002
18    16678.380705
24    14106.331371
25    15147.853723
30    13982.569521
Name: days_employed, dtype: float64

In [27]:
#переводим медиану дней трудового стажа из дней в годы

df['days_employed'].median() / 365

6.015604182343139

In [28]:
# заменяем пропущенные значения столбца "дни трудового стажа" медианами
# проверяем, остались ли пропуски после замены пропущенных значений медианами

df.loc[df['days_employed'].isna(), 'days_employed'] = df['days_employed'].median()
display('Пропуски в трудовом стаже после:', df['days_employed'].isna().sum())
display(df[df['days_employed'].isna()].head()) 

'Пропуски в трудовом стаже после:'

0

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose]
Index: []

In [29]:
# считаем пропуски в столбце "общий доход" 

display('Пропуски в общем доходе до:', df['total_income'].isna().sum())
display(df[df['total_income'].isna()].head()) 

'Пропуски в общем доходе до:'

2165

children  days_employed  dob_years education  education_id  \
12         0    2195.695527         65   среднее             1   
26         0    2195.695527         41   среднее             1   
29         0    2195.695527         63   среднее             1   
41         0    2195.695527         50   среднее             1   
55         0    2195.695527         54   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   

    total_income                           purpose  
12           NaN                   сыграть свадьбу  
26           NaN                       образование  
29           NaN  строительство жилой недвижимости  
41           NaN  сделка с подержанным автомобилем  
55           NaN                   сыграть свадьбу

In [30]:
# заменяем пропущенные значения столбца "общий доход" медианами
# проверяем, остались ли пропуски после замены пропущенных значений медианами

df.loc[df['total_income'].isna(), 'total_income'] = df['total_income'].median()
display('Пропуски в общем доходе после:', df['total_income'].isna().sum())

'Пропуски в общем доходе после:'

0

**Вывод**

Совпало количество пропусков в столбцах **days_employed** и **total_income** -2174. Это чуть меньше 10% от общего числа заемщиков.
Возможно, часть замещиков брали кредиты, не предоставляя справку о доходах и копию трудовой книжки.
Поэтому сотрудники банка не заполнили сведения о сумме всех источников дохода (**total_income**) и трудовом стаже(**days_employed**). 

Удалять я эти данные не стала потому, что удалять данные можно только в том случае, если их меньше 1%. А тут 10% - многовато для удаления.

### **<font color='DarkSeaGreen'> Замена типа данных </font>**

Поясняю, почему выбрала именно **astype** для замены вещественного типа данных на целочисленный. **to_numeric** здесь не подходит потому, что он предназначен для перевода строковых типов в численный. 

Параметр **errors='ignore'** выбран для того, чтобы игнорировать ошибки, которые выдает программа при попытке привести строковый тип **(str)** в целочисленный **(int)**


In [31]:
# приводим данные к целочисленному типу

df['days_employed'] = df['days_employed'].astype(int)
df['total_income'] = df['total_income'].astype(int)
df.dtypes

children             int64
days_employed        int64
dob_years            int64
education           object
education_id         int64
family_status       object
family_status_id     int64
gender              object
income_type         object
debt                 int64
total_income         int64
purpose             object
dtype: object

**Вывод**

Заменили тип данных в общем доходе и количестве дней трудового стажа на целочисленный.

### **<font color='LightSeaGreen'> Обработка дубликатов </font>** 

***Задание:***


**Удалите дубликаты**:
1. Поясните, как выбирали метод для поиска и удаления дубликатов в данных;
2. Приведите возможные причины появления дубликатов;

Сначала смотрим, нужно ли приводить текстовые поля к нижнему регистру, выводим уникальные значения для каждого из столбцов. В столбце **education** было видно, что одни и те же значения были введены разными регистрами, например, "СРЕДНЕЕ" и "среднее". Для того, чтобы обнаружить больше дубликатов, нужно было привести значения столбцов к одному регистру потому, что если бы все осталось так же, как и было, то многие дубликаты не были бы удалены. Что касается остальных столбцов - там все значения указаны в нижнем регистре, так что их не трогаем. 


***df.duplicated().sum()*** - этой командой я решила посчитать сумму дубликатов. Причем ДО того, как я написала  
***df = df.drop_duplicates().reset_index(drop=True)***, он количество дубликатов показывал 71. 

***df.duplicated().sum()*** #здесь он показывал 71 дубликат - но это было до того, как я удалила дубликаты  
***df = df.drop_duplicates().reset_index(drop=True)*** #удаление дубликатов с перенумерацией. Оставлять старую нумерацию нам незачем
***df.duplicated().sum()*** #дубликатов 0


Для начала посмотрим на то, какие есть уникальные значения в датафрейме. Это может нам дать подсказку, по каким столбцам может быть больше всего дубликатов. Особенно это касается неявных дубликатов, которые могут появляться, например, от разных регистров.

In [32]:
# выводим на экран список уникальных значений в столбце "образование"

display(df['education'].unique()) 

array(['высшее', 'среднее', 'Среднее', 'СРЕДНЕЕ', 'ВЫСШЕЕ',
       'неоконченное высшее', 'начальное', 'Высшее',
       'НЕОКОНЧЕННОЕ ВЫСШЕЕ', 'Неоконченное высшее', 'НАЧАЛЬНОЕ',
       'Начальное', 'Ученая степень', 'УЧЕНАЯ СТЕПЕНЬ', 'ученая степень'],
      dtype=object)

In [33]:
# выводим на экран список уникальных значений в столбце "семейный статус"

display(df['family_status'].unique())

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'Не женат / не замужем'], dtype=object)

In [34]:
# выводим  на экран список уникальных значений в столбце "тип занятости"

display(df['income_type'].unique())

array(['сотрудник', 'пенсионер', 'компаньон', 'госслужащий',
       'безработный', 'предприниматель', 'студент', 'в декрете'],
      dtype=object)

In [35]:
# выводим на экран список уникальных значений в столбце "цель кредита"

display(df['purpose'].unique())

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

In [36]:
#посмотрим, какие дубликаты у нас есть и сколько их

df[df.duplicated(keep=False)].sort_values(by=['total_income', 'days_employed'])

children  days_employed  dob_years education  education_id  \
120           0           2195         46   среднее             1   
520           0           2195         35   среднее             1   
541           0           2195         57   среднее             1   
554           0           2195         60   среднее             1   
680           1           2195         30    высшее             0   
...         ...            ...        ...       ...           ...   
20702         0           2195         64   среднее             1   
21032         0           2195         60   среднее             1   
21132         0           2195         47   среднее             1   
21281         1           2195         30    высшее             0   
21415         0           2195         54   среднее             1   

          family_status  family_status_id gender  income_type  debt  \
120     женат / замужем                 0      F    сотрудник     0   
520    гражданский брак                 1      F    сотрудник     0   
541     женат / замужем                 0      F    сотрудник     0   
554     женат / замужем                 0      M    сотрудник     0   
680     женат / замужем                 0      F  госслужащий     0   
...                 ...               ...    ...          ...   ...   
20702   женат / замужем                 0      F    пенсионер     0   
21032   женат / замужем                 0      F    пенсионер     0   
21132   женат / замужем                 0      F    сотрудник     0   
21281   женат / замужем                 0      F    сотрудник     0   
21415   женат / замужем                 0      F    пенсионер     0   

       total_income                            purpose  
120          145017                 высшее образование  
520          145017                    сыграть свадьбу  
541          145017   сделка с подержанным автомобилем  
554          145017               покупка недвижимости  
680          145017            покупка жилья для семьи  
...             ...                                ...  
20702        145017         дополнительное образование  
21032        145017              заняться образованием  
21132        145017                       ремонт жилью  
21281        145017  покупка коммерческой недвижимости  
21415        145017                  операции с жильем  

[106 rows x 12 columns]

Причина дубликатов - данные были внесены в несколько разных баз данных. А поскольку выгрузка была из нескольких разных источников (это предположение уже несколько раз подтвердилось, я об этом писала выше), то это и привело к созданию дублей. Даже если бы выгрузка была из одной единой системы, но данные могли при выгрузке задвоиться. 

In [37]:
# удаляем дубликаты
# проверяем, осталились ли дубликаты после удаления

df = df.drop_duplicates().reset_index(drop=True)

df.duplicated().sum()

0

In [38]:
# приводим все значения столбца "образование" к нижнему регистру для того, чтобы обнаружить явные дубликаты и удалить их, 
# помня о том, что python чувствителен к регистру и он не считает явными дубликатами одинаковые тексты (слова, значения), 
# написанные разными регистрами
# решила привести к нижнему регистру только столбец "образование" потому,что это - единственный столбец, в котором были
# обнаружены значения, написанные в смешанном или верхнем регистре. Остальные столбцы приводить к нижнему регистру 
# нет смысла потому, что они и так в записаны исключительно в нижнем регистре

df['education'] = df['education'].str.lower()
display(df.head())

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0          14177         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

In [39]:
# считаем уникальные значения столбца "образование"

df['education'].value_counts()

среднее                15128
высшее                  5237
неоконченное высшее      741
начальное                282
ученая степень             6
Name: education, dtype: int64

In [40]:
# выводим на экран уникальные значения столбца "образование"

display(df['education'].unique())

array(['высшее', 'среднее', 'неоконченное высшее', 'начальное',
       'ученая степень'], dtype=object)

In [41]:
# считаем дубликаты во всем датафрейме.

df.duplicated().sum()

17

In [42]:
# удаляем дубликаты, делаем перенумерацию индексов после удаления дубликатов
# проверяем, остались ли дубликаты после удаления. Видим, что не осталось

df = df.drop_duplicates().reset_index(drop=True)
df.duplicated().sum()

0

**Вывод**

Выявили все явные и неявные дубликаты. Удалили.

### **<font color='DarkCyan'> Лемматизация </font>** 


В столбце **'purpose' (цель получения кредита)** есть по сути одинаковые цели с той только разницей, что написаны разными словами и формулировками. Для того, чтобы привести данные по целям кредита к единообразию, существует лемматизация - приведение слова к его изначальной словарной форме. 

In [43]:
unique_purposes = df['purpose'].value_counts().index.tolist()
print(df['purpose'].value_counts().to_frame())

                                        purpose
свадьба                                     790
на проведение свадьбы                       764
сыграть свадьбу                             761
операции с недвижимостью                    674
покупка коммерческой недвижимости           658
покупка жилья для сдачи                     650
операции с жильем                           648
операции с коммерческой недвижимостью       646
покупка жилья                               643
жилье                                       642
покупка жилья для семьи                     637
недвижимость                                632
строительство собственной недвижимости      629
операции со своей недвижимостью             626
строительство жилой недвижимости            623
покупка своего жилья                        620
строительство недвижимости                  619
покупка недвижимости                        618
ремонт жилью                                605
покупка жилой недвижимости              

In [44]:
# считаем самые распространенные цели кредита.

m = Mystem()
list_of_lemmas = []
for element in df['purpose']:
    lemma = m.lemmatize(element)
    list_of_lemmas.extend(lemma)

unique_lemmas = Counter(list_of_lemmas)
sorted(unique_lemmas.items(), key = lambda pair: pair[1], reverse=True)

[(' ', 33445),
 ('\n', 21377),
 ('недвижимость', 6329),
 ('покупка', 5877),
 ('жилье', 4445),
 ('автомобиль', 4290),
 ('образование', 3998),
 ('с', 2905),
 ('операция', 2594),
 ('свадьба', 2315),
 ('свой', 2226),
 ('на', 2211),
 ('строительство', 1871),
 ('высокий', 1368),
 ('получение', 1311),
 ('коммерческий', 1304),
 ('для', 1287),
 ('жилой', 1227),
 ('сделка', 937),
 ('заниматься', 904),
 ('дополнительный', 902),
 ('проведение', 764),
 ('сыграть', 761),
 ('сдача', 650),
 ('семья', 637),
 ('собственный', 629),
 ('со', 626),
 ('ремонт', 605),
 ('подержанный', 482),
 ('подержать', 472),
 ('приобретение', 461),
 ('профильный', 434)]

Итак, возьмем 10 самых распространенных целей кредита. Поскольку команда unique_lemmas.head(10) не сработала (при помощи нее я хотела вывести топ-10 целей кредитов), то единственный вариант взять эти цели - вручную скопировать из списка выше. Так и сделаю.

недвижимость', 6329),  
 ('покупка', 5877),  
 ('жилье', 4445),  
 ('автомобиль', 4290),  
 ('образование', 3998),  
 ('с', 2905),  
 ('операция', 2594),  
 ('свадьба', 2315),  
 ('свой', 2226),  
 ('на', 2211)  

Из этого списка придется убрать:  

**'покупка'** - слишком широкое понятие. Цель кредита должна быть более конкретной. 

**'жилье'** - то же самое, что и "недвижимость". Синонимы.  

**'операция'**  - как мы уже видели в списке выше, имеется в виду не только медицинская операция, но и "операция с недвижимостью". Слишком двусмысленно.  

**'с'** - это явно не цель кредита. Просто распространенный предлог, часто встречающийся в формулировках целей кредита, не более того. 

**'свой'** - то же, что и пункт выше.  

**'на'** - то же, что и пункт выше  

**Оставляем:**

недвижимость'  
'автомобиль'  
'образование'  
'свадьба'  

Возьмем еще цели для того, чтобы набрать топ-10 целей (пока что мы набрали только 4, отсеялось 6). Берем побольше потому, что у нас идет большой отсев тех лемм, которые точно не являются целями кредита.

строительство', 1871),  
 ('высокий', 1368),  
 ('получение', 1311),  
 ('коммерческий', 1304),  
 ('для', 1287),  
 ('жилой', 1227),  
 ('сделка', 937),  
 ('заниматься', 904),  
 ('дополнительный', 902),  
 ('проведение', 764),  
 ('сыграть', 761),  
 ('сдача', 650),  
 ('семья', 637),  
 ('собственный', 629),  
 ('со', 626),  
 ('ремонт', 605)  

**строительство' - тоже относится к недвижимости**  
 ('высокий', 1368),  - явно не цель кредита  
 ('получение', 1311),  - нет конкретики в цели кредита  
 ('коммерческий', 1304),  - нет конкретики в цели кредита  
 ('для', 1287),  - нет конкретики в цели кредита  
 **('жилой', 1227),   - тоже относится к недвижимости**    
 ('сделка', 937),  - нет конкретики в цели кредита. Сделка, я так понимаю, купли-продажи. Вот только что покупали?   
 ('заниматься', 904),  - нет конкретики в цели кредита  
 ('дополнительный', 902),  - нет конкретики в цели кредита  
 ('проведение', 764),  - нет конкретики в цели кредита  
 ('сыграть', 761),  - нет конкретики в цели кредита  
 **('сдача', 650),  - тоже относится к недвижимости**  
 ('семья', 637),  - нет конкретики в цели кредита  
 ('собственный', 629),  - нет конкретики в цели кредита  
 ('со', 626),  - нет конкретики в цели кредита  
 **('ремонт', 605)  - тоже относится к недвижимости**   
 подержанный', 482),  - нет конкретики в цели кредита
 ('подержать', 472),  - нет конкретики в цели кредита
 ('приобретение', 461),  - нет конкретики в цели кредита
 ('профильный', 434)] - нет конкретики в цели кредита

В итоге набрали мы **5 уникальных целей кредита** из всего списка уникальных лемм, который у нас был:  

недвижимость'  
'автомобиль'  
'образование'  
'свадьба'    
'ремонт'

In [45]:
# создаем функцию, которая проверяет каждую ячейку в столбце purpose и присваивает соответствующую категорию

def categorize(purpose): 
    purpose_lemmas = m.lemmatize(purpose) 
    if 'недвижимость' in purpose_lemmas or 'жилье' in purpose_lemmas or 'строительство' in purpose_lemmas or 'жилой' in purpose_lemmas or 'сдача' in purpose_lemmas or 'ремонт' in purpose_lemmas:
        return 'недвижимость'
    if 'автомобиль' in purpose_lemmas:
        return 'автомобиль'
    if 'образование' in purpose_lemmas:
        return 'образование'
    if 'свадьба' in purpose_lemmas:
        return 'свадьба'
    return 'прочее'

df['category'] = df['purpose'].apply(categorize)


display(df.loc[:, ['purpose', 'category']])
display(df['category'])

purpose      category
0                     покупка жилья  недвижимость
1           приобретение автомобиля    автомобиль
2                     покупка жилья  недвижимость
3        дополнительное образование   образование
4                   сыграть свадьбу       свадьба
...                             ...           ...
21372             операции с жильем  недвижимость
21373          сделка с автомобилем    автомобиль
21374                  недвижимость  недвижимость
21375  на покупку своего автомобиля    автомобиль
21376         на покупку автомобиля    автомобиль

[21377 rows x 2 columns]

0        недвижимость
1          автомобиль
2        недвижимость
3         образование
4             свадьба
             ...     
21372    недвижимость
21373      автомобиль
21374    недвижимость
21375      автомобиль
21376      автомобиль
Name: category, Length: 21377, dtype: object

In [46]:
df['category'].value_counts()

недвижимость    10774
автомобиль       4290
образование      3998
свадьба          2315
Name: category, dtype: int64

**Вывод**

Лемматизировав цели кредита, привели данные к единообразию. Нам это может пригодиться в том случае, если нам нужно не только узнать, какие бывают самые распространенные цели кредита, но и посчитать их количество (см.ячейку выше). Например, мы можем сделать вывод о том, что на недвижимость приходится практически половина всех кредитов. На автомобиль или образование - каждый пятый кредит. На свадьбы берут каждый десятый кредит.

### **<font color='DodgerBlue'> Категоризация данных </font>** 

In [47]:
# считаем количество уникальных значений в столбце "дети"

df['children'].unique() 

array([1, 0, 3, 2, 4, 5])

In [48]:
# считаем количество уникальных значений в столбце "долг"

df['debt'].unique()

array([0, 1])

In [49]:
# считаем количество уникальных значений в столбце "ID семейного статуса"

df['family_status_id'].unique()

array([0, 1, 2, 3, 4])

In [50]:
# считаем количество уникальных значений в столбце "семейный статус"

df['family_status'].unique()

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'Не женат / не замужем'], dtype=object)

In [51]:
# выводим на экран среднее по доходу всех заемщиков

df['total_income'].mean()

165221.34710202555

In [52]:
# сводная таблица, которая показывает проценты должников в зависимости от кол-ва детей

pivt_children_debt = df.pivot_table(index = 'children', columns = 'debt', 
                                     values = 'total_income', aggfunc = 'count')

pivt_children_debt.set_axis(['debt_no', 'debt_yes'], axis = 1, inplace = True)

pivt_children_debt['percents'] = (pivt_children_debt['debt_yes'] / 
                                (pivt_children_debt['debt_yes'] + pivt_children_debt['debt_no']))

pivt_children_debt.sort_values(by = 'percents', ascending = False)

pivt_children_debt.style.format({'percents':'{:.1%}'})

display(pivt_children_debt.style.format({'percents':'{:.1%}'}))

pivt_children_debt = pivt_children_debt.sort_values(by = 'percents', ascending = False)
display(pivt_children_debt)

debt_no  debt_yes  percents
children                             
4            37.0       4.0  0.097561
2          1858.0     194.0  0.094542
1          4364.0     444.0  0.092346
3           303.0      27.0  0.081818
0         13073.0    1064.0  0.075263
5             9.0       NaN       NaN

In [53]:
# выводим сводную таблицу, которая показывает связь семейного положения и процентов просроченных кредитов

report_family_status = df.groupby('family_status').agg({'debt': ['mean', 'count']})
report_family_status.columns = ['% невозврата', 'Кол-во заемщиков']
report_family_status = report_family_status.sort_values(by = '% невозврата', ascending = False)
report_family_status.style.format({'% невозврата': '{:.2%}'})

In [54]:
report_children = df.groupby('children').agg({'debt': ['mean', 'count']})
report_children.columns = ['% невозврата', 'Кол-во заемщиков']
report_children = report_children.sort_values(by = '% невозврата', ascending = False)
report_children.style.format({'% невозврата': '{:.2%}'})

In [55]:
# выводим среднее по общему доходу для того, чтобы понять, какой доход является средним и
# в дальнейшем использовать эту цифру для категоризации данных по доходу

df['total_income'].mean()

165221.34710202555

In [56]:
# выводим медиану по общему доходу 

df['total_income'].median()

145017.0

In [57]:
# функция категоризации данных. Сделаем категоризацию по общему доходу и количеству детей. 
# по семейному статусу и цели кредита у нас категоризация уже есть

def total_income_cat(row):
    if row['total_income'] <= 50000:
        return 'низкий доход'
    elif 50000 < row['total_income'] <= 165000:
        return 'средний доход'
    elif 165000 < row['total_income'] < 1000000:
        return 'высокий доход'
    else:
        return 'миллионер'
  
    
def children_cat(row):
    if row['children'] == 0:
        return 'нет детей'
    elif 1 <= row['children'] <= 2:
        return '1-2 ребенка'
    else:
        return 'многодетные'

In [58]:
df['total_income_cat'] = df.apply(total_income_cat, axis=1)
df['children_cat'] = df.apply(children_cat, axis=1)

In [59]:
df.head()

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0          14177         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose      category total_income_cat children_cat  
0               покупка жилья  недвижимость    высокий доход  1-2 ребенка  
1     приобретение автомобиля    автомобиль    средний доход  1-2 ребенка  
2               покупка жилья  недвижимость    средний доход    нет детей  
3  дополнительное образование   образование    высокий доход  многодетные  
4             сыграть свадьбу       свадьба    средний доход    нет детей

## **<font color='CornflowerBlue'> Ответы на вопросы </font>** 

In [60]:
def func_cat(category):
    return df.groupby(category)['debt'].mean().to_frame().sort_values(by='debt')

- ***Есть ли зависимость между наличием детей и возвратом кредита в срок?***

In [61]:
func_cat('children_cat')

debt
children_cat          
нет детей     0.075263
многодетные   0.081579
1-2 ребенка   0.093003

**Вывод**

Да, зависимость есть. Среди замщиков, не вернувших кредит в срок, меньше всего бездетных - их 7,5%.   

Что касается остальных 93,5% заемщиков с просрочками по оплате, то у них дети есть. Зависимости от количества детей нет.

- ***Есть ли зависимость между семейным положением и возвратом кредита в срок?***

In [62]:
func_cat('family_status')

debt
family_status                  
вдовец / вдова         0.065969
в разводе              0.070411
женат / замужем        0.075509
гражданский брак       0.093040
Не женат / не замужем  0.097465

**Вывод**

**Самый большой процент просроченных кредитов** - среди заемщиков, не состоящих в браке (9,75%).  
После них по количеству просрочек идут те, кто состоит в гражданском браке (9,3%) и официально женатые/замужние (7,55%).   

**Самый маленький процент должников** - среди овдовевших (6,6 %) и среди разведенных (7,04%). 

- ***Есть ли зависимость между уровнем дохода и возвратом кредита в срок?***

In [63]:
func_cat('total_income_cat')

debt
total_income_cat          
низкий доход      0.061828
высокий доход     0.077596
миллионер         0.080000
средний доход     0.083602

**Вывод**

**Больше всего просрочек** по платежам - среди замещиков с **высоким** (7,7 %) и **средним** (8,3%) доходом.  
**Меньше всего** - с **низким** доходом (6,1 %).

- ***Как разные цели кредита влияют на его возврат в срок?***

In [64]:
func_cat('category')

debt
category              
недвижимость  0.072397
свадьба       0.079050
образование   0.092296
автомобиль    0.093473

**Вывод**

Меньше всего просрочек - у заемщиков, чья цель кредита так или иначе связана с **недвижимостью** (7,2 %). После них с небольшим отрывом идут заемщики, которые берут кредит на **свадьбу** (7,9 %).  

Больше всего должников с просроченными платежами - тех, кому нужны деньги на получение **образования** (9,2 %) и на приобретения **автомобиля** (9,3 %). 

## **<font color='MediumSlateBlue'> Общий вывод </font>**  <a id='4'></a>

Достигнута цель нашего исследования - **выяснить, влияет ли семейное положение и количество детей на погашение кредита в срок**.

Итак, мы приходим к выводам, что: 

**Самые ответственные заемщики - это**:

-по семейному положению - **разведенные** (7,04%) или **овдовевшие** (6,6 %)  
-по количеству детей - **бездетные** (7,5%)  
-по уровню дохода - **с низким доходом** (6,1 %)  
-по цели кредита - **недвижимость** (7,2 %) (и все, что с ней связано, например, ремонт, сделка купли-продажи, и т.д.) или **свадьба**  (7,9 %).



**Наименее ответственные заемщики - это:**

-по семейному положению - живущие в **гражданском браке** (9,3%), либо **не состоящие в браке** (9,75%)  
-по количеству детей - **имеющие детей (93%)**. Корелляции между количеством детей и просрочками по платежам не обнаружено.  
-по уровню дохода - с **высоким** (7,7 %) или **средним** (8,3%) **доходом**.  

-по цели кредита - **получение образования** (9,2 %) или **приобретение автомобиля** (9,3 %). 






## **<font color='RoyalBlue'> Чек-лист готовности проекта </font>** 

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [x] файл изучен;
- [x]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x]  есть пояснение, какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [x]  объяснено, по какому принципу заполнены пропуски;
- [x]  заменен вещественный тип данных на целочисленный;
- [x]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [x]  удалены дубликаты;
- [x]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [x]  описаны возможные причины появления дубликатов в данных;
- [x]  выделены леммы в значениях столбца с целями получения кредита;
- [x]  описан процесс лемматизации;
- [x]  данные категоризированы;
- [x]  есть объяснение принципа категоризации данных;
- [x]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [x]  в каждом этапе есть выводы;
- [x]  есть общий вывод.